File: C/acquire_interrupt_split_trigger_thread.cpp
/* Red Pitaya C++ API example for acquiring data from all 4 channels in split
 * trigger mode with per-channel threads and multiple acquisitions */

#include "rp.h"
#include "rp_hw-profiles.h"
#include "rp_hw_calib.h"
#include <algorithm>
#include <chrono>
#include <iomanip>
#include <iostream>
#include <mutex>
#include <random>
#include <sstream>
#include <thread>
#include <vector>

#define MAX_NUM 4
#define BUFF_SIZE 6001
#define TIMEOUT_MS 3000
#define ACQUISITIONS_PER_CHANNEL 5 // Number of acquisitions per channel

using namespace std;

/* ============================================================ */
/* CHANNEL CONFIGURATION STRUCTURE */
/* ============================================================ */

struct ChannelConfig {
  int channel_num;           // Channel number: 1, 2, 3, 4
  rp_channel_t ch;           // Red Pitaya channel handle
  rp_channel_trigger_t trig; // Trigger channel
  uint32_t decimation;       // Decimation factor
  float trig_lvl;            // Trigger level in Volts
  int trig_dly;              // Trigger delay
  rp_acq_trig_src_t src;     // Trigger source
  uint32_t trig_pos;         // Trigger position in buffer
  vector<float> data;        // Acquired data buffer
  int acquisition_count;     // Number of successful acquisitions

  ChannelConfig() : channel_num(0), trig_pos(0), acquisition_count(0) {
    data.resize(BUFF_SIZE);
  }
};

/* ============================================================ */
/* RED PITAYA INITIALIZATION */
/* ============================================================ */

/**
 * Initialize Red Pitaya hardware and enable split trigger mode
 * Initialization ends with rp_AcqSetSplitTrigger(true)
 * Returns true on success, false on failure
 */
bool initRedPitaya() {
  if (rp_InitReset(false) != RP_OK) {
    cerr << "RP init failed!" << endl;
    return false;
  }

  if (rp_CalibInit() != RP_HW_CALIB_OK) {
    cerr << "Calibration failed!" << endl;
    return false;
  }

  rp_AcqReset();
  rp_AcqSetSplitTrigger(true); // <-- INITIALIZATION ENDS HERE

  return true;
}

/* ============================================================ */
/* THREAD-SAFE PRINT FUNCTION */
/* ============================================================ */

/**
 * Thread-safe print function using variadic templates
 * Automatically locks mutex and appends newline
 */
template <typename... Args> void safePrint(mutex &cout_mutex, Args... args) {
  lock_guard<mutex> lock(cout_mutex);
  ((cout << args), ...); // C++17 fold expression
  cout << endl;
}

/* ============================================================ */
/* PER-CHANNEL THREAD FUNCTION */
/* ============================================================ */

/**
 * Thread function for a single channel
 * Each thread performs: configuration -> multiple acquisitions -> output
 */
void channelThread(ChannelConfig &cfg, int thread_id, mutex &cout_mutex) {
  // ============================================================
  // STEP 1: CHANNEL CONFIGURATION (done once)
  // ============================================================

  safePrint(cout_mutex, "[THREAD ", thread_id, "] Configuring channel ",
            cfg.channel_num);

  // Disable interrupt for trigger point
  rp_AcqSetIntMaskCh(cfg.ch, RP_INT_TRIGGER, false);

  // Reset channel
  rp_AcqResetCh(cfg.ch);

  // Set decimation factor
  rp_AcqSetDecimationFactorCh(cfg.ch, cfg.decimation);

  // Set gain to low range
  rp_AcqSetGain(cfg.ch, RP_LOW);

  // Set trigger level
  rp_AcqSetTriggerLevel(cfg.trig, cfg.trig_lvl);

  // Set trigger delay
  rp_AcqSetTriggerDelayCh(cfg.ch, cfg.trig_dly);

  safePrint(cout_mutex, "[THREAD ", thread_id, "] Channel ", cfg.channel_num,
            " configured");

  // ============================================================
  // STEP 2: MULTIPLE DATA ACQUISITIONS
  // ============================================================

  for (int acq = 0; acq < ACQUISITIONS_PER_CHANNEL; acq++) {
    // --------------------------------------------------------
    // 2a. START ACQUISITION AND SET TRIGGER SOURCE
    // --------------------------------------------------------

    // Start acquisition on channel
    rp_AcqStartCh(cfg.ch);

    // Set trigger source for channel
    rp_AcqSetTriggerSrcCh(cfg.ch, cfg.src);

    // --------------------------------------------------------
    // 2b. WAIT FOR TRIGGER
    // --------------------------------------------------------

    safePrint(cout_mutex, "[THREAD ", thread_id, "] Acquisition ", acq + 1,
              ": Waiting for trigger on channel ", cfg.channel_num);

    // Wait for trigger and buffer full (timeout in ms)
    auto ret = rp_AcqIntFillReadCh(cfg.ch, TIMEOUT_MS);

    // Check if trigger was successful
    if (ret != RP_OK) {
      safePrint(cout_mutex, "[THREAD ", thread_id, "] Acquisition ", acq + 1,
                ": Channel ", cfg.channel_num,
                " trigger failed! Error: ", rp_GetError(ret));
      continue; // Skip this acquisition attempt
    }

    safePrint(cout_mutex, "[THREAD ", thread_id, "] Acquisition ", acq + 1,
              ": Channel ", cfg.channel_num, " triggered, ret = ", ret);

    // Check if trigger was successful
    if (ret != RP_OK) {
      safePrint(cout_mutex, "[THREAD ", thread_id, "] Acquisition ", acq + 1,
                ": Channel ", cfg.channel_num, " trigger failed!");
      continue; // Skip this acquisition attempt
    }

    // --------------------------------------------------------
    // 2c. ACQUIRE DATA
    // --------------------------------------------------------

    safePrint(cout_mutex, "[THREAD ", thread_id, "] Acquisition ", acq + 1,
              ": Acquiring data from channel ", cfg.channel_num);

    // Get trigger position in buffer
    rp_AcqGetWritePointerAtTrigCh(cfg.ch, &cfg.trig_pos);

    safePrint(cout_mutex, "[THREAD ", thread_id, "] Acquisition ", acq + 1,
              ": Channel ", cfg.channel_num, " trig pos = ", cfg.trig_pos);

    // Read data from buffer starting at trigger position
    uint32_t size = BUFF_SIZE;
    rp_AcqGetDataV(cfg.ch, cfg.trig_pos, &size, cfg.data.data());

    // Increment acquisition counter
    cfg.acquisition_count++;

    safePrint(cout_mutex, "[THREAD ", thread_id, "] Acquisition ", acq + 1,
              ": Channel ", cfg.channel_num, " data acquired (", size,
              " samples)");

    // --------------------------------------------------------
    // 2d. OUTPUT DATA
    // --------------------------------------------------------

    {
      lock_guard<mutex> lock(cout_mutex);
      cout << "[THREAD " << thread_id << "] Acquisition " << acq + 1
           << ": Channel " << cfg.channel_num << " data:" << endl;

      // Print only first 20 samples for brevity
      int print_samples = min(20, (int)cfg.data.size());
      cout << fixed << setprecision(3);
      for (int j = 0; j < print_samples; j++) {
        cout << cfg.data[j];
        if (j < print_samples - 1)
          cout << ", ";
      }
      cout << endl << endl;
    }

    // Small delay between acquisitions to allow hardware to settle
    this_thread::sleep_for(chrono::milliseconds(100));
  }

  // ============================================================
  // STEP 3: COMPLETION SUMMARY
  // ============================================================

  safePrint(cout_mutex, "[THREAD ", thread_id, "] Channel ", cfg.channel_num,
            " completed ", cfg.acquisition_count, " acquisitions");
}

/* ============================================================ */
/* MAIN FUNCTION */
/* ============================================================ */

int main() {
  // Get number of available ADC channels
  int ch_num = rp_HPGetFastADCChannelsCountOrDefault();

  // Initialize Red Pitaya hardware
  if (!initRedPitaya()) {
    return -1;
  }

  // ============================================================
  // CHANNEL CONFIGURATION SETUP
  // ============================================================

  // Channel hardware mappings
  rp_channel_t ch_list[MAX_NUM] = {RP_CH_1, RP_CH_2, RP_CH_3, RP_CH_4};
  rp_channel_trigger_t trig_list[MAX_NUM] = {RP_T_CH_1, RP_T_CH_2, RP_T_CH_3,
                                             RP_T_CH_4};
  rp_acq_trig_src_t src_list[MAX_NUM] = {
      RP_TRIG_SRC_CHA_PE, // Channel A positive edge
      RP_TRIG_SRC_CHB_PE, // Channel B positive edge
      RP_TRIG_SRC_CHC_PE, // Channel C positive edge
      RP_TRIG_SRC_CHD_PE  // Channel D positive edge
  };

  // Default configuration values
  uint32_t dec_list[MAX_NUM] = {64, 64, 64, 64};
  float lvl_list[MAX_NUM] = {0, 0, 0, 0};
  int dly_list[MAX_NUM] = {0, 0, 0, 0};

  // Create channel configurations
  vector<ChannelConfig> channels(ch_num);
  for (int i = 0; i < ch_num; i++) {
    channels[i].channel_num = i + 1;
    channels[i].ch = ch_list[i];
    channels[i].trig = trig_list[i];
    channels[i].decimation = dec_list[i];
    channels[i].trig_lvl = lvl_list[i];
    channels[i].trig_dly = dly_list[i];
    channels[i].src = src_list[i];
  }

  // Mutex for thread-safe console output
  mutex cout_mutex;

  // Print initial configuration info
  safePrint(cout_mutex, "Acquisitions per channel: ", ACQUISITIONS_PER_CHANNEL);
  safePrint(cout_mutex, "");

  // ============================================================
  // START PER-CHANNEL THREADS
  // ============================================================

  vector<thread> threads;
  for (int i = 0; i < ch_num; i++) {
    threads.emplace_back(channelThread, ref(channels[i]), i + 1,
                         ref(cout_mutex));
  }

  // Wait for all threads to complete
  for (auto &t : threads) {
    t.join();
  }

  // ============================================================
  // FINAL SUMMARY
  // ============================================================

  safePrint(cout_mutex, "");
  safePrint(cout_mutex, "[MAIN] ===== SUMMARY =====");

  int total_acquisitions = 0;
  for (auto &cfg : channels) {
    safePrint(cout_mutex, "[MAIN] Channel ", cfg.channel_num, ": ",
              cfg.acquisition_count, " acquisitions");
    total_acquisitions += cfg.acquisition_count;
  }
  safePrint(cout_mutex, "[MAIN] Total acquisitions: ", total_acquisitions);

  // ============================================================
  // CLEANUP
  // ============================================================

  rp_Release();

  return 0;
}

File: python-api/acquire_interrupt_split_trigger_thread.py
#!/usr/bin/python3
"""
Red Pitaya Python API example for acquiring data from all 4 channels in split
trigger mode with per-channel threads and multiple acquisitions.
"""

import rp
import rp_hw_profiles
import rp_hw_calib
import threading
import time
import numpy as np
from typing import List

# Constants
MAX_NUM = 4
BUFF_SIZE = 6001
TIMEOUT_MS = 3000
ACQUISITIONS_PER_CHANNEL = 5  # Number of acquisitions per channel


class ChannelConfig:
    """Channel configuration structure."""

    def __init__(self):
        self.channel_num = 0  # Channel number: 1, 2, 3, 4
        self.ch = None  # Red Pitaya channel handle
        self.trig = None  # Trigger channel
        self.decimation = 0  # Decimation factor
        self.trig_lvl = 0.0  # Trigger level in Volts
        self.trig_dly = 0  # Trigger delay
        self.src = None  # Trigger source
        self.trig_pos = 0  # Trigger position in buffer
        self.data = np.zeros(BUFF_SIZE, dtype=np.float32)  # Acquired data buffer as numpy array
        self.acquisition_count = 0  # Number of successful acquisitions


# ============================================================
# RED PITAYA INITIALIZATION
# ============================================================

def init_red_pitaya():
    """
    Initialize Red Pitaya hardware and enable split trigger mode.
    Initialization ends with rp_AcqSetSplitTrigger(true).
    Returns True on success, False on failure.
    """
    # Reset Red Pitaya
    if rp.rp_InitReset(False) != rp.RP_OK:
        print("RP init failed!", flush=True)
        return False

    # Initialize calibration
    if rp_hw_calib.rp_CalibInit() != rp_hw_calib.RP_HW_CALIB_OK:
        print("Calibration failed!", flush=True)
        return False

    # Reset acquisition and set split trigger mode
    rp.rp_AcqReset()
    rp.rp_AcqSetSplitTrigger(True)  # <-- INITIALIZATION ENDS HERE

    return True


# ============================================================
# THREAD-SAFE PRINT FUNCTION
# ============================================================

def safe_print(cout_lock: threading.Lock, *args):
    """Thread-safe print function that automatically locks mutex and appends newline."""
    with cout_lock:
        print(*args, flush=True)


# ============================================================
# PER-CHANNEL THREAD FUNCTION
# ============================================================

def channel_thread(cfg: ChannelConfig, thread_id: int, cout_lock: threading.Lock):
    """
    Thread function for a single channel.
    Each thread performs: configuration -> multiple acquisitions -> output.
    """

    # ============================================================
    # STEP 1: CHANNEL CONFIGURATION (done once)
    # ============================================================

    safe_print(cout_lock, f"[THREAD {thread_id}] Configuring channel {cfg.channel_num}")

    # Disable interrupt for trigger point
    rp.rp_AcqSetIntMaskCh(cfg.ch, rp.RP_INT_TRIGGER, False)

    # Reset channel
    rp.rp_AcqResetCh(cfg.ch)

    # Set decimation factor
    rp.rp_AcqSetDecimationFactorCh(cfg.ch, cfg.decimation)

    # Set gain to low range
    rp.rp_AcqSetGain(cfg.ch, rp.RP_LOW)

    # Set trigger level
    rp.rp_AcqSetTriggerLevel(cfg.trig, cfg.trig_lvl)

    # Set trigger delay
    rp.rp_AcqSetTriggerDelayCh(cfg.ch, cfg.trig_dly)

    safe_print(cout_lock, f"[THREAD {thread_id}] Channel {cfg.channel_num} configured")

    # ============================================================
    # STEP 2: MULTIPLE DATA ACQUISITIONS
    # ============================================================

    for acq in range(ACQUISITIONS_PER_CHANNEL):
        # --------------------------------------------------------
        # 2a. START ACQUISITION AND SET TRIGGER SOURCE
        # --------------------------------------------------------

        # Start acquisition on channel
        rp.rp_AcqStartCh(cfg.ch)

        # Set trigger source for channel
        rp.rp_AcqSetTriggerSrcCh(cfg.ch, cfg.src)

        # --------------------------------------------------------
        # 2b. WAIT FOR TRIGGER
        # --------------------------------------------------------

        safe_print(
            cout_lock,
            f"[THREAD {thread_id}] Acquisition {acq + 1}: "
            f"Waiting for trigger on channel {cfg.channel_num}"
        )

        # Wait for trigger and buffer full (timeout in ms)
        ret = rp.rp_AcqIntFillReadCh(cfg.ch, TIMEOUT_MS)

        # Check if trigger was successful
        if ret != rp.RP_OK:
            safe_print(
                cout_lock,
                f"[THREAD {thread_id}] Acquisition {acq + 1}: "
                f"Channel {cfg.channel_num} trigger failed! "
                f"Error: {rp.rp_GetError(ret)}"
            )
            continue  # Skip this acquisition attempt

        safe_print(
            cout_lock,
            f"[THREAD {thread_id}] Acquisition {acq + 1}: "
            f"Channel {cfg.channel_num} triggered, ret = {ret}"
        )

        # --------------------------------------------------------
        # 2c. ACQUIRE DATA
        # --------------------------------------------------------

        safe_print(
            cout_lock,
            f"[THREAD {thread_id}] Acquisition {acq + 1}: "
            f"Acquiring data from channel {cfg.channel_num}"
        )

        # Get trigger position in buffer
        result, cfg.trig_pos = rp.rp_AcqGetWritePointerAtTrigCh(cfg.ch)

        safe_print(
            cout_lock,
            f"[THREAD {thread_id}] Acquisition {acq + 1}: "
            f"Channel {cfg.channel_num} trig pos = {cfg.trig_pos}"
        )

        # Read data from buffer using rp_AcqGetDataVNP with numpy array
        ret = rp.rp_AcqGetDataVNP(cfg.ch, cfg.trig_pos, cfg.data)

        if ret != rp.RP_OK:
            safe_print(
                cout_lock,
                f"[THREAD {thread_id}] Acquisition {acq + 1}: "
                f"Channel {cfg.channel_num} data read failed! "
                f"Error: {rp.rp_GetError(ret)}"
            )
            continue  # Skip this acquisition attempt

        # Increment acquisition counter
        cfg.acquisition_count += 1

        safe_print(
            cout_lock,
            f"[THREAD {thread_id}] Acquisition {acq + 1}: "
            f"Channel {cfg.channel_num} data acquired ({len(cfg.data)} samples)"
        )

        # --------------------------------------------------------
        # 2d. OUTPUT DATA
        # --------------------------------------------------------

        with cout_lock:
            print(
                f"[THREAD {thread_id}] Acquisition {acq + 1}: "
                f"Channel {cfg.channel_num} data:",
                flush=True
            )

            # Print only first 20 samples for brevity
            print_samples = min(20, len(cfg.data))
            formatted_data = ", ".join(f"{cfg.data[j]:.3f}" for j in range(print_samples))
            print(formatted_data, flush=True)
            print(flush=True)

        # Small delay between acquisitions to allow hardware to settle
        time.sleep(0.1)  # 100 ms

    # ============================================================
    # STEP 3: COMPLETION SUMMARY
    # ============================================================

    safe_print(
        cout_lock,
        f"[THREAD {thread_id}] Channel {cfg.channel_num} "
        f"completed {cfg.acquisition_count} acquisitions"
    )


# ============================================================
# MAIN FUNCTION
# ============================================================

def main():
    """Main function to orchestrate multi-channel acquisition."""

    # Get number of available ADC channels
    ch_num = rp_hw_profiles.rp_HPGetFastADCChannelsCountOrDefault()

    # Initialize Red Pitaya hardware
    if not init_red_pitaya():
        return -1

    # ============================================================
    # CHANNEL CONFIGURATION SETUP
    # ============================================================

    # Channel hardware mappings
    ch_list = [rp.RP_CH_1, rp.RP_CH_2, rp.RP_CH_3, rp.RP_CH_4]
    trig_list = [rp.RP_T_CH_1, rp.RP_T_CH_2, rp.RP_T_CH_3, rp.RP_T_CH_4]
    src_list = [
        rp.RP_TRIG_SRC_CHA_PE,  # Channel A positive edge
        rp.RP_TRIG_SRC_CHB_PE,  # Channel B positive edge
        rp.RP_TRIG_SRC_CHC_PE,  # Channel C positive edge
        rp.RP_TRIG_SRC_CHD_PE   # Channel D positive edge
    ]

    # Default configuration values
    dec_list = [64, 64, 64, 64]
    lvl_list = [0, 0, 0, 0]
    dly_list = [0, 0, 0, 0]

    # Create channel configurations
    channels: List[ChannelConfig] = []
    for i in range(ch_num):
        cfg = ChannelConfig()
        cfg.channel_num = i + 1
        cfg.ch = ch_list[i]
        cfg.trig = trig_list[i]
        cfg.decimation = dec_list[i]
        cfg.trig_lvl = lvl_list[i]
        cfg.trig_dly = dly_list[i]
        cfg.src = src_list[i]
        channels.append(cfg)

    # Lock for thread-safe console output
    cout_lock = threading.Lock()

    # Print initial configuration info
    safe_print(cout_lock, f"Acquisitions per channel: {ACQUISITIONS_PER_CHANNEL}")
    safe_print(cout_lock, "")

    # ============================================================
    # START PER-CHANNEL THREADS
    # ============================================================

    threads = []
    for i in range(ch_num):
        thread = threading.Thread(
            target=channel_thread,
            args=(channels[i], i + 1, cout_lock)
        )
        threads.append(thread)
        thread.start()

    # Wait for all threads to complete
    for thread in threads:
        thread.join()

    # ============================================================
    # FINAL SUMMARY
    # ============================================================

    safe_print(cout_lock, "")
    safe_print(cout_lock, "[MAIN] ===== SUMMARY =====")

    total_acquisitions = 0
    for cfg in channels:
        safe_print(
            cout_lock,
            f"[MAIN] Channel {cfg.channel_num}: "
            f"{cfg.acquisition_count} acquisitions"
        )
        total_acquisitions += cfg.acquisition_count

    safe_print(cout_lock, f"[MAIN] Total acquisitions: {total_acquisitions}")

    # ============================================================
    # CLEANUP
    # ============================================================

    rp.rp_Release()

    return 0


if __name__ == "__main__":
    exit(main())